llvm/test/Transforms/Float2Int/basic.ll

```diff
@@ -349,3 +349,55 @@ bogusBB:                                          ; preds = %bogusBB
   %tobool = fcmp une double %inc, 0.000000e+00
   br label %bogusBB
 }
+
+define i32 @pr79158() {
+; CHECK-LABEL: @pr79158(
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    [[X_I:%.*]] = alloca i32, align 4
+; CHECK-NEXT:    call void @llvm.lifetime.start.p0(i64 4, ptr nonnull [[X_I]])
+; CHECK-NEXT:    store volatile i32 1, ptr [[X_I]], align 4
+; CHECK-NEXT:    [[X_I_0_X_I_0_X_0_X_0_X_0__I:%.*]] = load volatile i32, ptr [[X_I]], align 4
+; CHECK-NEXT:    [[CMP_I:%.*]] = icmp sgt i32 [[X_I_0_X_I_0_X_0_X_0_X_0__I]], 0
+; CHECK-NEXT:    [[TMP0:%.*]] = zext i1 [[CMP_I]] to i32
+; CHECK-NEXT:    [[MUL_I1:%.*]] = mul i32 [[TMP0]], -1
+; CHECK-NEXT:    call void @llvm.lifetime.end.p0(i64 4, ptr nonnull [[X_I]])
+; CHECK-NEXT:    ret i32 0
+;
+entry:
+  %x.i = alloca i32, align 4
+  call void @llvm.lifetime.start.p0(i64 4, ptr nonnull %x.i)
+  store volatile i32 1, ptr %x.i, align 4
+  %x.i.0.x.i.0.x.0.x.0.x.0..i = load volatile i32, ptr %x.i, align 4
+  %cmp.i = icmp sgt i32 %x.i.0.x.i.0.x.0.x.0.x.0..i, 0
+  %conv.i = uitofp i1 %cmp.i to double
```

Comment:
The range in the debug print for this instruction seems problematic

```
F2I:   %conv.i = uitofp i1 %cmp.i to double:[0,1)
```

That range only contains one value, 0.


The pass computed a minimum bitwidth of 3

```
F2I: MinBitwidth=3, R: [0,1)
```

Reply:
> The range in the debug print for this instruction seems problematic
> 
> ```
> F2I:   %conv.i = uitofp i1 %cmp.i to double:[0,1)
> ```
> 
> That range only contains one value, 0.
> 
> The pass computed a minimum bitwidth of 3
> 
> ```
> F2I: MinBitwidth=3, R: [0,1)
> ```

Don't know if that is something for another PR because I only touched the conversion aspect, upon which the type converted to was set to 32 bits anyway

Reply:
Whether minbitwith is 3 or 1, it only matters in choosing a 32 or 64 bit type. At least, right now that is what happens.

Reply:
> The range in the debug print for this instruction seems problematic
> 
> ```
> F2I:   %conv.i = uitofp i1 %cmp.i to double:[0,1)
> ```
> 
> That range only contains one value, 0.
> 
> The pass computed a minimum bitwidth of 3
> 
> ```
> F2I: MinBitwidth=3, R: [0,1)
> ```

That means either R.getLower().getSignificantBits() or R.getHigher().getSignificantBits() returned 2.

Because that + 1 makes 3

Reply:
If the value is not negative, this
function returns the same value as getActiveBits()+1. 

Okay so assuming 1 has 1 active bit, that makes 2, and then + 1 is 3 for MinBitWidth

Reply:
The broken constant range is probably the real reason for the test failure. Your fix is only masking it.

Reply:
So, should it be 2?

Reply:
The correct range for that uitofp is [0, 2). I think the bug is in ConstantRange::castOp. I'm working on a fix.

Reply:
This fixes it

```
diff --git a/llvm/lib/IR/ConstantRange.cpp b/llvm/lib/IR/ConstantRange.cpp
index 3394a1ec8dc4..484740204d16 100644
--- a/llvm/lib/IR/ConstantRange.cpp
+++ b/llvm/lib/IR/ConstantRange.cpp
@@ -746,7 +746,7 @@ ConstantRange ConstantRange::castOp(Instruction::CastOps CastOp,
       Min = Min.zext(ResultBitWidth);
       Max = Max.zext(ResultBitWidth);
     }
-    return ConstantRange(std::move(Min), std::move(Max));
+    return ConstantRange(std::move(Min), Max + 1);
   }
   case Instruction::SIToFP: {
     // TODO: use input range if available
```

The second argument to the ConstantRange constructor is supposed to be first value outside the range. The variable `Max` is the last value inside the range so we need to increment it.

Reply:
Does this fix the miscompile as a whole too?

Reply:
With that change, the minimum bitwidth is calculate as 35 so we pick i64 as the type. Which produces this

```
Args: ./bin/opt test.ll -passes=float2int -S -o - -debug 
F2I: Looking at function pr79158
F2I:   %conv1.i = fptoui double %mul.i to i32:empty-set
F2I:   %mul.i = fmul double %conv.i, 0x41EFFFFFFFE00000:empty-set
F2I:   %conv.i = uitofp i1 %cmp.i to double:[0,2)
F2I:   %mul.i = fmul double %conv.i, 0x41EFFFFFFFE00000:[0,4294967296)
F2I:   %conv1.i = fptoui double %mul.i to i32:[0,4294967296)
F2I: MinBitwidth=35, R: [0,4294967296)
; ModuleID = 'test.ll'
source_filename = "test.ll"

define i32 @pr79158() {
entry:
  %x.i = alloca i32, align 4
  call void @llvm.lifetime.start.p0(i64 4, ptr nonnull %x.i)
  store volatile i32 1, ptr %x.i, align 4
  %x.i.0.x.i.0.x.0.x.0.x.0..i = load volatile i32, ptr %x.i, align 4
  %cmp.i = icmp sgt i32 %x.i.0.x.i.0.x.0.x.0.x.0..i, 0
  %0 = zext i1 %cmp.i to i64
  %mul.i1 = mul i64 %0, 4294967295
  %1 = trunc i64 %mul.i1 to i32
  call void @llvm.lifetime.end.p0(i64 4, ptr nonnull %x.i)
  ret i32 0
}
```

Reply:
Perhaps I should remove the + 1 from the MinBW assigned to it from the max of the range ends

Reply:
> Perhaps I should remove the + 1 from the MinBW assigned to it from the max of the range ends

Why?

Reply:
34 makes more sense than 35 for UINT_MAX

Reply:
> 34 makes more sense than 35 for UINT_MAX

How so?

Reply:
define i32 @pr79158_2() {
; CHECK-LABEL: @pr79158_2(
; CHECK-NEXT:  entry:
; CHECK-NEXT:    [[X_I:%.*]] = alloca i32, align 4
; CHECK-NEXT:    call void @llvm.lifetime.start.p0(i64 4, ptr nonnull [[X_I]])
; CHECK-NEXT:    store volatile i32 1, ptr [[X_I]], align 4
; CHECK-NEXT:    [[X_I_0_X_I_0_X_0_X_0_X_0__I:%.*]] = load volatile i32, ptr [[X_I]], align 4
; CHECK-NEXT:    [[CMP_I:%.*]] = icmp sgt i32 [[X_I_0_X_I_0_X_0_X_0_X_0__I]], 0
; CHECK-NEXT:    [[TMP0:%.*]] = zext i1 [[CMP_I]] to i32
; CHECK-NEXT:    [[MUL_I1:%.*]] = mul i32 [[TMP0]], -2147483648
; CHECK-NEXT:    call void @llvm.lifetime.end.p0(i64 4, ptr nonnull [[X_I]])
; CHECK-NEXT:    ret i32 0

I feel like i32 should be chosen in particular I mean.